File: src/SIL.Machine.TS/src/statistics/log-space.ts
export const LOG_ONE: number = 0;
export const LOG_ZERO: number = -999999999;

export function logAdd(logx: number, logy: number): number {
  if (logx > logy) {
    return logx + Math.log(1 + Math.exp(logy - logx));
  }
  return logy + Math.log(1 + Math.exp(logx - logy));
}

export function logMultiply(logx: number, logy: number): number {
  let result = logx + logy;
  if (result < LOG_ZERO) {
    result = LOG_ZERO;
  }
  return result;
}

export function logDivide(logx: number, logy: number): number {
  let result = logx - logy;
  if (result < LOG_ZERO) {
    result = LOG_ZERO;
  }
  return result;
}

export function toLogSpace(value: number): number {
  return Math.log(value);
}

export function toStandardSpace(logValue: number): number {
  return Math.exp(logValue);
}

File: src/SIL.Machine/Statistics/LogSpace.cs
using System;

namespace SIL.Machine.Statistics
{
    public static class LogSpace
    {
        public const double One = 0;
        public const double Zero = -999999999;

        public static double Add(double logx, double logy)
        {
            if (logx > logy)
                return logx + Math.Log(1 + Math.Exp(logy - logx));
            return logy + Math.Log(1 + Math.Exp(logx - logy));
        }

        public static double Multiply(double logx, double logy)
        {
            double result = logx + logy;
            if (result < Zero)
                result = Zero;
            return result;
        }

        public static double Divide(double logx, double logy)
        {
            double result = logx - logy;
            if (result < Zero)
                result = Zero;
            return result;
        }

        public static double ToLogSpace(double value)
        {
            return Math.Log(value);
        }

        public static double ToStandardSpace(double logvalue)
        {
            return Math.Exp(logvalue);
        }
    }
}
